internal/parser/parser.go

```diff
@@ -0,0 +1,157 @@
+package parser
+
+import (
+	"fmt"
+	"go/types"
+
+	"golang.org/x/tools/go/packages"
+	"golang.org/x/tools/go/types/typeutil"
+	"golang.org/x/tools/imports"
+
+	"github.com/Makpoc/gomigen/internal/model"
+)
+
+// Parse loads the package under dir and looks for an interface with the given name.
+// If one is found Parse parses its components to build a model.Middleware.
+func Parse(dir, interfaceName string) (*model.Middleware, error) {
+	targetPackage, targetInterface, err := loadInterface(dir, interfaceName)
+	if err != nil {
+		return nil, err
+	}
+
+	middleware := model.NewMiddleware(interfaceName)
+
+	originalInterfacePkgAlias := middleware.Imports.Add(model.Import{
+		Alias:   targetPackage.Name,
+		Package: targetPackage.PkgPath,
+	})
+	middleware.OriginalPackage = model.Package{
+		Name: originalInterfacePkgAlias,
+		Path: targetPackage.PkgPath,
+	}
+
+	middleware.GenPackage = targetPackage.Name + "mw"
+
+	mSet := typeutil.IntuitiveMethodSet(targetInterface, nil)
+	for _, m := range mSet {
+		err = processMethod(middleware, m)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse method: %w", err)
+		}
+	}
+	return middleware, nil
+}
+
+func loadInterface(dir string, interfaceName string) (pkg *packages.Package, iface *types.Named, err error) {
+	p, err := packages.Load(&packages.Config{
+		Mode: packages.NeedName | packages.NeedImports | packages.NeedTypes,
+	}, dir)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to load packages in directory %q: %w", dir, err)
+	}
+	if len(p) == 0 {
+		return nil, nil, fmt.Errorf("no go packages found in directory %q", dir)
+	}
+
+	targetPackage := p[0]
+	pkgScope := targetPackage.Types.Scope()
+	targetIFaceObj := pkgScope.Lookup(interfaceName)
+	if targetIFaceObj == nil {
+		return nil, nil, fmt.Errorf("no interface named %s in package %s", interfaceName, targetPackage.PkgPath)
+	}
+	targetIFace, ok := (targetIFaceObj.Type()).(*types.Named)
+	if !ok {
+		return nil, nil, fmt.Errorf("%s is not an interface declaration", interfaceName)
+	}
+
+	return targetPackage, targetIFace, nil
+}
+
+func processMethod(middleware *model.Middleware, method *types.Selection) error {
+	if method == nil || method.Obj() == nil || method.Type() == nil {
+		return fmt.Errorf("unexpected method definition: %+v", method)
+	}
+
+	f, ok := method.Obj().(*types.Func)
+	if !ok {
+		// for go 1.17 interface this should not happen.
+		// TODO test with go 1.18 and interface with generic constraints.
+		return nil
```

Comment:
Do you want to be ignoring this error? If you don't process this method, won't the final middleware be invalid? Isn't it better to return an error here than to make the user guess what is going on.

Reply:
I've changed it to return an error, similar to the check for types.Signature just below it.